### PR TITLE
servlet: Implement gRPC server as a Servlet

### PR DESCRIPTION
**Pull request by niloc132**
This is an updated version of the code found at https://github.com/grpc/grpc-java/pull/4738, merged up to latest, tests restored from [a different branch](https://github.com/dapengzhang0/grpc-java/commit/d1aaafcac), and with a few of the remaining review comments addressed as well. 

The tests run during the build against Jetty 9, 10, 11, Tomcat 9, 10, and Undertow 9. Note that Jetty 10 and 11 are largely the same except for the javax/jakarta package changes). There are still several tests marked as FIXME, and several more ignored due to how servlets function differently than something entirely self-contained like grpc-netty. Note also that while Jetty 9 works, it requires a workaround for missing support for writing trailers - I am amenable to dropping this workaround or documenting it in some way for downstream users, but without this Jetty 9 can't be tested in Java 8. Note also that a standard Jetty 9 installation will not support this use case, but it either needs to be lightly customized, or used as an embedded web server.

A second project is also added, grpc-servlet-jakarta, which consists of build logic to rewrite package names for `javax.servlet` imports and replace them with their corresponding `jakarta.servlet` imports. This also removes the Jetty workaround, as it is unnecessary (Jetty 11+ supports the `jakarta.servlet` APIs, which necessarily means that trailers are supported using the standard API calls).

--

As a new contributor to this project, I anticipate that I have made some mistakes in trying to bring my parts of this patch up to the standards expected for contribution, and it is quite possible that I missed some conversation where it was decided that this wouldn't make sense to merge to grpc-java itself. If that is the case, I will be releasing this separately - outside of tests, there should be no other changes to the modules in grpc-java itself. With this released one way or the other, my intention is to add grpc-web support here as well, to allow a standalone (i.e. without Envoy) Java web server to provide static html/js content and grpc services, which otherwise requires ~3 processes (and potentially one more to support non-SSL streaming to the browser).

Inspecting my own code, I think it makes sense to improve usability of the tests in IntelliJ, and add README.md files to each of grpc-servlet and grpc-servlet-jakarta.

### CODE REVIEW

File: servlet/src/jettyTest/java/io/grpc/servlet/JettyInteropTest.java
Comment (dapengzhang0) [inline servlet/src/jettyTest/java/io/grpc/servlet/JettyInteropTest.java:66]: (I'm not familiar with Jetty) What's the reason to add rate control?

**Review by dapengzhang0 — COMMENTED** [review]
Thank you @niloc132  for your great effort for merging it up to date and adding the Jakarta support. Still reviewing, just one comment and a couple of questions for the moment.

File: servlet/src/test/java/io/grpc/servlet/TomcatInteropTest.java
Comment (niloc132) [reply servlet/src/test/java/io/grpc/servlet/TomcatInteropTest.java]: I found this to be necessary, for the same reason as the jetty rate control below - otherwise, the "busy" grpc streams tripped some heuristic where the server thought it was being attacked, and it responded with an error rather than complete the test method. I'll add some documentation here and to the patch.

**Comment by niloc132** [comment]
> Could it be done completely reflectively? I really don't want to have strong dependency on container-specific classes (e.g. org.eclipse.jetty.server.*) here.

(I can't seem to reply to your message here, so making a general reply instead.)

Perhaps, though you'll notice that this class only has a compile-time dependency on Jetty, and won't require a runtime dependency unless Jetty is already in use. This was written this way to avoid the need for explicit reflection, so as to make it easier to understand, but I'm of the opinion at this point that it isn't worth the trouble. My self-review pointed this out too, jetty 9 might just be more trouble than it is worth. 

I had hoped to keep Jetty 9.4 support, as it seems to have better http2 support than the other two servlet containers you had written tests for, and it still supports Java 8, but the jetty project isn't interested in patching this bug as part of their release, and there are a few ways that this workaround can still go wrong, so I'm inclined to remove it entirely, except perhaps as a footnote to enable embedded jetty 9.4 support.

File: servlet/src/test/java/io/grpc/servlet/TomcatInteropTest.java
Comment (niloc132) [reply servlet/src/test/java/io/grpc/servlet/TomcatInteropTest.java]: https://stackoverflow.com/a/61454503/860630

> The overhead protection was put in place in response to a collection of CVE's reported against HTTP/2 in the middle of 2019. While Tomcat wasn't directly affected (the malicious input didn't trigger excessive load) we did take steps to block input that matched the malicious profile.
...
> The overhead count starts at -10. For every DATA frame received it is reduced by 1. For every SETTINGS, PRIORITY and PING frame it is increased by overheadCountFactor.If the overhead count goes above 0, the connection is closed.
> In addition, if the average size of a received non-final DATA frame and the previously received DATA frame (on that same stream) is less than overheadDataThreshold then the overhead count is increased by overheadDataThreshold/(average size of current and previous DATA frames). In this way, the smaller the DATA frame, the greater the increase in the overhead. A small number of small non-final DATA frames should be enough to trigger connection closure.

https://tomcat.apache.org/tomcat-9.0-doc/config/http2.html has more information about the various "overhead*" configuration keys and their roles, though `overheadDataThreshold`'s documentation says that _non-final_ data frames increase the count rather than reduce it. This appear to be reflected in code, see tomcat's Http2UpgradeHandler.startRequestBodyFrame:
https://github.com/apache/tomcat/blob/9.0.x/java/org/apache/coyote/http2/Http2UpgradeHandler.java#L1510-L1531
```
        // DATA frames reduce the overhead count ...
        reduceOverheadCount(FrameType.DATA);

        // .. but lots of small payloads are inefficient so that will increase
        // the overhead count unless it is the final DATA frame where small
        // payloads are expected.

        // See also https://bz.apache.org/bugzilla/show_bug.cgi?id=63690
        // The buffering behaviour of some clients means that small data frames
        // are much more frequent (roughly 1 in 20) than expected. Use an
        // average over two frames to avoid false positives.
        if (!endOfStream) {
          //...
```

--

Assuming that we didn't want to fine tune these settings for the sake of tests (which are deliberately pathological in some ways, to confirm the behavior of the implementation). Downstream users would need to correctly configure their servlet container to account for their expected usage vs concern over this class of attack.

File: servlet/src/jettyTest/java/io/grpc/servlet/JettyInteropTest.java
Comment (niloc132) [reply servlet/src/jettyTest/java/io/grpc/servlet/JettyInteropTest.java:66]: As with tomcat, jetty defaults to some degree of protection here, and also needs to be turned down. Jetty 9 has no errors, but in Jetty 10, the veryLargeResponse test fails with the server canceling the stream in the JettyInteropTest.

**Comment by dapengzhang0** [comment]
Sorry didn't get a chance to focus on this in the past weeks, will find time in December.

**Comment by dapengzhang0** [comment]
@niloc132, I pushed some small fixes to your branch.

File: settings.gradle
Comment (dapengzhang0) [reply settings.gradle]: Oh, this is outdated.

Comment (niloc132) [reply settings.gradle]: No - I abandoned this approach and made my own by hand in the -jakarta build. It isn't working right at the moment, I'll put up a patch that fixes it.

File: servlet/src/tomcatTest/java/io/grpc/servlet/TomcatInteropTest.java
Comment (niloc132) [reply servlet/src/tomcatTest/java/io/grpc/servlet/TomcatInteropTest.java:136]: At one point it failed intermittently, and then it apparently stopped failing? I haven't had time to figure out why it fails, left the test running so that the next time there might be a chance to figure it out.

**Comment by niloc132** [comment]
We've copied these changes [directly into the repo](https://github.com/deephaven/deephaven-core/tree/main/grpc-java) I'm working on, with one minor alteration so that a servlet-based grpc-web adapter can also use the same ServerTransportListener etc. This is certainly the wrong way to approach this problem, but I don't yet have the right one - I'll come back to the grpc-servlet with a followup as we finish implementing the other pieces we need on top, but my hunch is that the GrpcServlet will mostly be replaced with a GrpcFilter that handles application/grpc calls, and lets the others pass through to other appropriate servlets/filters.

**Comment by niloc132** [comment]
Jetty9 workaround and tests have been removed, and the jakarta build correctly publishes an artifact now as well.

I'm still struggling with several tomcat tests that intermittently fail, I'll spend a little time on them over the next few days, but will probably err on the side of ignoring them (to be revisited as the tested tomcat version is updated) - does that seem reasonable?

Thanks for taking the time to review this, I appreciate it.

**Comment by dapengzhang0** [comment]
>Jetty9 workaround and tests have been removed, and the jakarta build correctly publishes an artifact now as well.

Thank you so much!


> I'm still struggling with several tomcat tests that intermittently fail, I'll spend a little time on them over the next few days, but will probably err on the side of ignoring them (to be revisited as the tested tomcat version is updated) - does that seem reasonable?

Yes, if it's not easy too fix/triage, for the moment we will ignore them and I'll file a tracking issue later, and investigate in the future. In the past, I filed a number of tomcat bugs/enhancement related to async servlet over HTTP/2, some of the intermittent failures took me a lot of time to find a minimal reproducible verifiable example, it was quite painful.

**Comment by niloc132** [comment]
I've pushed another commit where i set another three flags to instruct tomcat to disregard other cases where the client is too chatty for its likes. Several dozen test iterations later, I feel confident in reporting that none of the three intermittent tests are failing any longer.

File: servlet/jakarta/build.gradle
Comment (dapengzhang0) [inline servlet/jakarta/build.gradle:40]: Can you also migrate `io/grpc/servlet/*.java` to  something like `io/grpc/servletjakarta/*.java`? Different APIs with exactly the same class and package name will potentially cause trouble.

Comment (niloc132) [reply servlet/jakarta/build.gradle:40]: Yes I'll get that resolved this week, sorry for the delay.

Comment (dapengzhang0) [reply servlet/jakarta/build.gradle:40]: Thanks. BTW, there are many possibilities for the package naming like, but not limited to, the following. As a user, just take whatever is most convenient for you.

| Java EE 8      | Jakarta EE |
| ----------- | ----------- |
| io.grpc.servlet |  io.grpc.servletjakarta |
| io.grpc.servlet |  io.grpc.servlet.jakarta |
| io.grpc.servlet.javaee8   | io.grpc.servlet |
| io.grpc.servlet.javaee8 | io.grpc.servlet.jakarta |
| io.grpc.servlet.javax | io.grpc.servlet.jakarta |

Comment (niloc132) [reply servlet/jakarta/build.gradle:40]: I went with `io.grpc.servlet` -> `io.grpc.servlet.jakarta`.

**Comment by niloc132** [comment]
There were warnings from java 9-16 on tomcat's illegal reflective access on startup in the tests, I added the appropriate `--add-opens` and validated that tests pass cleanly on Java 11 and 17.

File: examples/example-servlet/src/main/java/io/grpc/servlet/examples/helloworld/HelloWorldServlet.java
Comment (dapengzhang0) [reply examples/example-servlet/src/main/java/io/grpc/servlet/examples/helloworld/HelloWorldServlet.java:36]: I agree we should have an example to use extension of `GrpcServlet` directly, maybe implementing the [routeguide](https://github.com/grpc/grpc-java/tree/v1.44.0/examples/src/main/java/io/grpc/examples/routeguide) service, in addition to the existing example.

As for Filter, that might be something we can add after this PR being merged, and have this PR only support the most basic servlet API. Same for the grpc-web support.

**Comment by niloc132** [comment]
Is there anything I can do to help get this into the right shape to merge and release it?

**Comment by dapengzhang0** [comment]
> Is there anything I can do to help get this into the right shape to merge and release it?

@niloc132 It's almost in the right shape now. I'm trying to add some unit tests. I'm actively on it.

**Comment by niloc132** [comment]
Thanks - if there is something you'd like me to help out on, leave a note and I'll get to it as soon as possible. 

We're doing some experimenting with grpc-web in another ServletAdapter-like class - this makes more sense if you can share the transportlistener, streamtracerfactories etc between them, but I don't have a pattern I like well enough yet to propose here. For the moment at least, we'll maintain our own fork as we iterate this, and propose something when we have some options to suggest, see what fits best into the rest of the project here.

We're also trying using a single Filter to handle all mapped services rather than bind a servlet to each path or bind a specific servlet for each service, so that unhandled paths (or non-grpc calls) can just fall through to other filters or servlets. This also makes it straightforward to create separate handlers for gRPC calls based on the same transportlisterner/streamtracerfactory/etc, as for example websockets* can't be configured consistently across servlet containers to start from a filter/servlet. 

* Websockets of course are not a standard grpc/grpc-web transport mechanism, but https://github.com/improbable-eng/grpc-web/ offers a handy client transport that allows a browser to connect to streams with binary data through a websocket proxy. This has also made localhost web development easier, as http2 doesn't work in a browser without tls. I don't imagine that the grpc-java project will be interested in the in-process Java implementation of that transport that we've developed, but we are going to make it generally available for use cases such as ours.

**Review by dapengzhang0 — APPROVED** [review]
LGTM

- The tests seem to be stable now. I do notice there's still some flakyness but the servlet module seems less flaky than some known flakiness in other modules. If some tests turn out to be more flaky than expected and annoying after PR is merged, we can `@Ignore` them later.

- There are some potential improvement, such as adding an example using `GrpcServlet` directly unlike the current one using `ServletAdapter`, worth doing but can be done after this PR.

- The `AsyncServletOutputStreamWriter` is still using a lock-free approach when calling isReady() to avoid deadlock with servlet container (https://github.com/apache/tomcat/blob/10.0.0/java/org/apache/coyote/Response.java#L744, https://github.com/apache/tomcat/blob/10.0.0/java/org/apache/coyote/Response.java#L771). This makes the writer class less readible than the using-lock approach though. This also makes it hard to see its concurrency correctness but a lincheck is added to justify it. This class has been thoroughly tested with three major servlet containers over the years and no issues caused by the writer's logic were found.

- While some more features such as Servelet Filter are requested/proposed by users, I would prefer to merge what we have now, make the basic feature available to use first, and users can contribute based on it.

@ejona86 PTAL

**Comment by dapengzhang0** [comment]
@ejona86 Any bandwidth for the final review? It's becoming more and more difficult to make the PR up-to-date as there could be regressions in the main branch that's only impacting grpc-servlet such as #9308 . 

@niloc132, I could not push new commits to this PR anymore. Did you disallow edit by the maintainers?

**Comment by niloc132** [comment]
The checkbox for edits by maintainers is still checked, and I made no deliberate change on my end. Push to a branch on your repo and I'll pull the commits over directly?

**Comment by niloc132** [comment]
Done - hopefully updating this will let this work again for you to push, otherwise just ping me and I'll update it again.

**Comment by hypnoce** [comment]
Thanks for all this effort. Is there anything blocking the merge (besides the failed kokro jobs) ? I'm really interested in this development as it can really simplify operations for us by sharing the same servlet container, same port, for both standard HTTP/REST calls and GRPC calls.

**Comment by hypnoce** [comment]
@niloc132 thanks a lot for your link. We'll definitively look at them. Unfortunately we use javax.* in our code and changing this will be difficult. @dapengzhang0 anything missing in this PR ?

**Comment by niloc132** [comment]
@hypnoce given that we're coming up on a year of this PR having been updated, I'm going to be moving forward with publishing a complete version (i.e. `javax.servlet` as well as `jakarta.servlet`) in the next month or two, plus the additional enhancements (grpc-web support, websocket transports, etc) that we've developed. I'll update the linked ticket and this pull request with that information once I've finished this process.

**Comment by rufreakde** [comment]
> @hypnoce given that we're coming up on a year of this PR having been updated, I'm going to be moving forward with publishing a complete version (i.e. `javax.servlet` as well as `jakarta.servlet`) in the next month or two, plus the additional enhancements (grpc-web support, websocket transports, etc) that we've developed. I'll update the linked ticket and this pull request with that information once I've finished this process.

@niloc132 
Will you update this ticket here then as well? Any news?

**Comment by niloc132** [comment]
@rufreakde 
>  Will you update this ticket here then as well? Any news?
Yes, when I get the chance to complete this work, I'll update the ticket and this PR. I'm sorry that I havent had time to do so, but as mentioned above, we've shipped the `jakarta.servlet` jars to maven central already.

File: servlet/src/jettyTest/java/io/grpc/servlet/JettyTransportTest.java
Comment (rufreakde) [reply servlet/src/jettyTest/java/io/grpc/servlet/JettyTransportTest.java:228]:
```suggestion
  @Ignore("This doesn't apply: Ensure that for a closed ServerStream, interactions are NoOps")
```
maybe noops as well

**Comment by niloc132** [comment]
@larry-safran Thank you for the review, I'll try to follow up shortly to make those changes. Can you confirm that make those will lead us down the path of getting this merged? I'm a little hesitant to invest further at this point.

**Comment by larry-safran** [comment]
Yes.  We can merge if none of the tests are unexpectedly failing and will fix them if they are.

**Comment by ejona86** [comment]
FYI, larry-safran was taking a look because I was saying I wanted to get this over the finish line. This has not been merged for far too long; it'd serve the code and everyone better to get it merged. For the small nits and fixes, I am completely fine with doing them in follow-ups or us maintainers pushing the fixes directly to this PR. I'm also fine with disabling a servlet-specific test or three (like #9308) in order to get it in.

(:-/, #9790 looks to have noticed new style failures which are failing the CI.)

**Comment by niloc132** [comment]
Thank you - I don't mean to sound pessimistic, and will get these done ASAP. I'll have a few followups to propose that we've already made too, but didn't want to further burden this PR with.

File: servlet/src/jettyTest/java/io/grpc/servlet/JettyTransportTest.java
Comment (dapengzhang0) [reply servlet/src/jettyTest/java/io/grpc/servlet/JettyTransportTest.java:246]: I filed #9308 to track the issue (which can be referenced here in the annotation). Seems there's no update to the issue. One can bump to the latest grpc version to see if the regression is still there. Otherwise need to do a bisect from v1.45 (the annotation mentioned v1.46 but actually it was bumping from v1.45) to v1.47 to find which change has caused the issue.

Comment (niloc132) [reply servlet/src/jettyTest/java/io/grpc/servlet/JettyTransportTest.java:246]: I've merged up to latest and the test still fails, incremented the version.

**Review by larry-safran — APPROVED** [review]

**Comment by larry-safran** [comment]
@niloc132 Please go ahead and merge this.

**Comment by ejona86** [comment]
@larry-safran, they don't have permission to merge this. For non-maintainer contributions, once we get two approvals we do the merge (typically the one who gives the second approval).

(Make sure to clean up the commit message when squashing.)

**Pull request merged**
